src/environment.ts

```diff
@@ -167,82 +158,96 @@ export class SecureEnvironment {
-                        // setter, and probably getter branch
-                        rawDescriptor = getSecureGlobalAccessorDescriptor(this, rawDescriptor);
-                    } else if (hasOwnProperty(rawDescriptor, 'get')) {
-                        // getter only branch (e.g.: window.navigator)
-                        const { get: originalGetter } = rawDescriptor;
-                        const env = this;
-                        rawDescriptor.get = function get(): SecureValue {
-                            const value: RawValue = apply(originalGetter as () => any, env.getRawValue(this), emptyArray);
-                            return env.getSecureValue(value);
-                        };
-                    } else {
-                        // value branch
+            // TODO: this whole loop needs cleanup and simplification avoid
+            // overriding ECMA script global keys.
+            if (ESGlobalKeys.has(key) || !hasOwnProperty(rawDescriptors, key)) {
+                continue;
+            }
+
+            // avoid poisoning to only installing own properties from baseDescriptors
+            const rawDescriptor = assign({}, rawDescriptors[key]);
+            setPrototypeOf(rawDescriptor, null);
+            if ('value' in rawDescriptor) {
+                // TODO: maybe we should make everything a getter/setter that way
+                // we don't pay the cost of creating the proxy in the first place
+                rawDescriptor.value = this.getSecureValue(rawDescriptor.value);
+            } else {
+               // Use the original getter to return a secure object, but if the 
+               // sandbox attempts to set it to a new value, this mutation will
+               // only affect the sandbox's global object, and the getter will
+               // start returning the new provided value rather than calling onto
+               // the outer realm. This is to preserve the object graph of the
+               // outer realm.
+                const env = this;
+                const { get: originalGetter } = rawDescriptor;
+                
+                let currentGetter = !isFunction(rawDescriptor.get) ? () => undefined : function(this: any): SecureValue {
+                    const value: RawValue = apply(originalGetter as () => any, env.getRawValue(this), emptyArray);
+                    return env.getSecureValue(value);
+                };
+
+                rawDescriptor.get = function(): SecureValue {
+                    return apply(currentGetter, this, emptyArray);
+                };
+
+                if (isFunction(rawDescriptor.set)) {
+                    rawDescriptor.set = function(v: SecureValue): void {
+                        // if a global setter is invoke, the value will be use as it is as the result of the getter operation
+                        currentGetter = () => v;
+                    };
+                }
+            }
```

Comment:
☝️ Simplified the logic here, we were duplicating the `getter` logic inside the `getSecureGlobalAccessorDescriptor()` helper.